Updating for new TOC structure

choldgraf opened the pull request:
This PR updates the Table of Contents documentation according to our new TOC structure (from #1293). I updated the content in the "Structure your Book" section, as well as in the "Getting Started" section. I also updated a few other content files so the structure made sense.

cc @chrisjsewell to make sure I haven't documented anything incorrectly, or that I haven't missed something...

**Note**: This is currently blocked on the items in #1323 - we need to make a release just after this is merged, so holding off on merging until we are ready to make a release.

# Tasks

- [x] Merge sphinx-external-toc update
- [x] Resolve https://github.com/executablebooks/jupyterbook-latex/issues/59
- [x] Bump sphinx-external-toc
- [x] Update sphinx-external-toc here
- [x] Release

choldgraf commented on docs/_toc.yml:
```diff
@@ -35,9 +40,10 @@ parts:
     - file: content/metadata
 
-  - file: execute/index
+  - file: content/executable/index
```
I'd like to re-organize these files so they follow a proper folder sub-structure, but we'll need to create redirects for this, and I don't wanna do it in this PR because it'll make the review a big pain

choldgraf commented on docs/content/code-outputs.md:
````diff
@@ -246,353 +246,3 @@ sphinx:
       - "text/plain"
 ```
-
-(content:code-outputs:glue)=
````
This is just me splitting this off into it's own page so that it is more discoverable

chrisjsewell commented on docs/structure/sections-headers.md:
```diff
@@ -0,0 +1,126 @@
+:::{tip}
+A good rule of thumb is to take one of these two approaches:
+
+1. **don't put headers in your introduction pages**. This is
```
Note that jupyterbook-latex essentially removes all headers in the introduction (aka root) document, maeaning that its child documents will all start at H2

choldgraf replied:
Interesting...do you think we should document this differently? Should it block this PR? (or if not, perhaps we should open an issue to clarify this point?)

mmcky replied:
Here is the `pdflatex` artifact that was build on the latest `master` re: behaviour of `jupyterbook-latex` around the front page (root document). 

https://github.com/executablebooks/jupyter-book/suites/2660999633/artifacts/58566503

It essentially treats the `root` page as `front-matter` (loosely)

chrisjsewell commented on docs/structure/configure.md:
````diff
@@ -0,0 +1,114 @@
+To add numbers to **all chapters of your book**, add the `numbered: true` flag to your book's defaults, like so:
+
+```yaml
+format: jb-book
````
you should mention `jb-article` as well

choldgraf replied:
In the latest commit I've added a mention of `jb-article`, lemme know how that looks

mmcky commented on docs/basics/page.md:
```diff
@@ -1,22 +1,49 @@
+format: jb-article
+root: index
+sections:
+- file: path/to/chapter1
```
@choldgraf wondering if this should be

```
- file: path/to/section1
```

rather than

```
- file: path/to/chapter1
```

I think that may reduce confusion. as you can and should use `jb-book` format for a chapter based book

```
format: jb-book
root: intro
chapters:
- file: chapter1
```

choldgraf replied:
yes definitely!

mmcky reviewed (commented):
@choldgraf nice work on these changes!

chrisjsewell commented on docs/structure/toc-generate.md:
````diff
@@ -0,0 +1,42 @@
+using the filenames of your book's content. To do so, run the following command
+
+```bash
+jupyter-book toc from-site path/to/book
````
Forgot to mention, here you need to specify the format you want, e.g. `-f jb-article` or `-f jb-book` then also you should explain the other options available (unless that is somewhere else)

choldgraf replied:
ah yeah - I added that option to the example - we show the `help` output at the bottom of the page, that work?

choldgraf commented:
OK I think I addressed the comment from @chrisjsewell and also added in a CHANGELOG entry. Looking at the PRs from last release, there wasn't actually a ton of movement on this repository, so the only major change was the `external-toc` stuff...if I missed something let me know and I'll add it to the CHANGELOG

chrisjsewell commented on CHANGELOG.md:
```diff
@@ -1,5 +1,25 @@
+
+([full changelog](https://github.com/executablebooks/jupyter-book/compare/v0.10.2...d43216c57a2710eaeb2363f21d2b3aadf914feca)) | [GitHub contributors page for this release](https://github.com/executablebooks/jupyter-book/graphs/contributors?from=2021-04-04&to=2021-05-11&type=c)
+
+This is primarily a documentation and under-the-hood update, but it comes with an important change to the way that **Table of Contents** are structured. See below for more information.
```
Ah you can get it more hype than this 😄. Maybe mention about the globbing feature, and that you can now use other file formats (basically mentioning some of the issues that the toc PR closed)
Also might want to specifically mention that the `jupyter-book toc` CLI command has changed to a group (and the original command is now equivalent to `jupyter-book toc from-site`)

choldgraf replied:
I agree - I'm just working on very limited bandwidth right now. I will try and get to this and flesh out the changelog a bit but it'll depend on baby / other work duties.

chrisjsewell replied:
Yeh no worries 👍

mmcky replied:
@choldgraf I will try and take a look at this today.

mmcky commented:
@choldgraf how do I update this branch that belongs to you? Can't seem to push to it. Accidentally setup another `toc-update` branch which I will delete. I have attached the `md` file I made changes to. Mainly just added a `NEW` section

[CHANGELOG.md](https://github.com/executablebooks/jupyter-book/files/6491260/CHANGELOG.md)

mmcky commented:
@chrisjsewell while working through some updates the CHANGELOG I added this section re: update to `jb toc` command:

````
- The `jupyterbook toc` command has now been converted to a collection of tools. 
  It was originally setup to read a collection of files and then build a `_toc.yml`. 
  The equivalent of this command is now:

  ```bash
  jupyter-book toc from-site
  ```

  There is also an additional helper tool that builds a collection of files from a user written `_toc.yml` file using

  ```bash
  jupyter-book toc to-site
  ```
````

I had one comment on the choice `to-site` and `from-site` -- this is very `html` centric. Should it be `to-project` and `from-project`?

choldgraf commented:
Ah that is a good point about site vs project, I think we should avoid web-specific terminology if possible. Or even "to-files" and "from-files" to be most explicit?

mmcky commented:
@choldgraf I think `to-files` and `from-files` is nice 👍

choldgraf commented:
@chrisjsewell what do you think about this pr? https://github.com/executablebooks/sphinx-external-toc/pull/34

choldgraf commented:
Thanks @mmcky for the updates, we're now mentioning `glob`, extra file formats, as well as some of the new TOC CLI features that @chrisjsewell mentioned.

I think that once https://github.com/executablebooks/sphinx-external-toc/pull/34 is merged and `sphinx-external-toc` releases a new version, we're ready to release a new Jupyter Book 🎉

choldgraf commented:
OK we've released a new `sphinx-external-toc` but hit a snag because of a circular dependency with `jupyterbook-latex`. I've opened up https://github.com/executablebooks/jupyterbook-latex/issues/59 to track resolving this. once it's fixed, we can bump the version of jupyterbook-latex and then cut a new release.

choldgraf commented:
OK I think that we are good to go here. I updated a couple of the commit messages so that we had a cleaner record of what changed in this PR, and I'm planning to merge via rebase to keep those commits discrete.

The one test is failing because codecov's servers aren't responding, I think it is just being flaky - in an earlier commit this worked fine, and some restarts make it work, others make it fail. All the test suites are passing, so I think we are OK.
